cmd/aries-agent-lambda/transport/apigw_ws/inbound.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/transport"
+	"github.com/hyperledger/aries-framework-go/pkg/framework/aries"
+)
+
+var (
+	transportProvider transport.Provider
+)
+
+type Inbound struct {
+	externalAddr string
+}
+
+// UnpackMessage using 'pack' with a 'source' of either 'ws' or 'http'.
+func UnpackMessage(message []byte, pack transport.Packager, source string) (*transport.Envelope, error) {
+	doubleQuote := []byte("\"")
+	msg := message
+
+	if bytes.HasPrefix(message, doubleQuote) && bytes.HasSuffix(message, doubleQuote) {
+		log.Printf("unpack msg from %s is wrapped with double quotes trying to base64 decode before unpacking..",
+			source)
+
+		msg = msg[1 : len(msg)-1]
+
+		var decodedMsg []byte
+
+		decodedMsg1, err1 := base64.URLEncoding.DecodeString(string(msg))
+		decodedMsg2, err2 := base64.RawURLEncoding.DecodeString(string(msg))
+
+		switch {
+		case err1 == nil:
+			decodedMsg = decodedMsg1
+		case err2 == nil:
+			decodedMsg = decodedMsg2
+		default:
+			return nil, fmt.Errorf("not base64 encoded message error from %s: URLEncoding error: %w, RawURLEncoding"+
+				" error: %v", source, err1, err2)
+		}
+
+		msg = decodedMsg
+	}
+
+	unpackMsg, err := pack.UnpackMessage(msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack msg from %s: %w", source, err)
+	}
+
+	return unpackMsg, nil
+}
+
+func NewInboundHandler() http.HandlerFunc {
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		processWSRequest(w, r, transportProvider)
+	})
+
+	return handler
+}
+
+func processWSRequest(w http.ResponseWriter, r *http.Request, prov transport.Provider) {
+	log.Printf("Processing WS request: %+v", r)
+	connectionId := r.Header.Get("X-ConnectionId")
+	log.Printf("Connection ID: %s", connectionId)
+	switch r.Method {
+	case "PUT":
+		{
+			// TODO: Websocket connected, save a reference to it in the DB
+			log.Printf(">>>>> Connect Event")
+			w.WriteHeader(http.StatusOK)
+		}
+	case "POST":
+		{
+
+			log.Printf(">>>>> WS Message Event")
+
+			message, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				log.Fatalf("read body failed : %v", err)
+				return
+			}
+
+			unpackMsg, err := UnpackMessage(message, transportProvider.Packager(), "ws")
+			if err != nil {
+				log.Fatalf("unpack message failed : %v", err)
+				return
+			}
+
+			trans := &decorator.Transport{}
+			err = json.Unmarshal(unpackMsg.Message, trans)
+			if err != nil {
+				log.Fatalf("unmarshal transport decorator : %v", err)
+				return
+			}
+
+			respChannel := make(chan []byte)
+			getConnPool().addKey(unpackMsg, trans, connectionId, respChannel)
+
+			err = prov.InboundMessageHandler()(unpackMsg)
+			if err != nil {
+				log.Fatalf("incoming msg processing failed: %v", err)
+				return
+			}
+
+			select {
+			//case response := <-respChannel:
+			//	w.WriteHeader(http.StatusOK)
+			//	log.Printf(">>>>>>>>> Got response\n")
+			//	w.Write(response)
+			case <-time.After(5 * time.Second):
```

Comment:
We needed this, otherwise auto-approve would never send the reply back (Lambda would terminate the app once the main function finished, and waiting on all goroutines wasn't feasible as some weren't meant to terminate).

Reply:
How are you writing response here?

Reply:
We just let the `transport/apigw_ws/outbound.go` handle the response writing (which is basically the same behavior as default WebSockets have). So technically we'd just need a way of getting notified that:
1. Yes, there's a reply to be sent now as part of this run, and
2. When that reply has actually been sent, so we can resume the flow here.